blockstack/blockstackd.py

```diff
@@ -3782,8 +3782,16 @@ class Program {
             name = {}
             name['name'] = name_info['name']
             name['address'] = b58ToC32(str(name_info['address']))
+
             name['expire_block'] = name_info['expire_block']
-            name['registered_at'] = max(name_info['first_registered'], name_info['last_renewed'])
+            if name['expire_block'] != -1: # for special flag `-1` do not subtract the block height
+                name['expire_block'] = name['expire_block'] - block
+
+            name['renewal_deadline'] = name_info['renewal_deadline']
+            if name['renewal_deadline'] != -1: # for special flag `-1` do not subtract the block height
+                name['renewal_deadline'] = name['renewal_deadline'] - block
+
+            name['registered_at'] = max(name_info['first_registered'], name_info['last_renewed']) - block
```

Comment:
This looks like it could be negative?  Isn't `block` the current chain height, in which case, `first_registered` and `last_renewed` may precede it?

Reply:
Yes this is intended, see PR description. It could be absolute value if we don't want to deal with signed integers, but I figured we'd keep all the "block delta" fields logically consistent. 

Any suggestions?

Reply:
Yeah I wonder if BNS block heights should be based on the burnchain block height, instead of the stacks chain block height.
Any opinion @jcnelson @kantai ?

Reply:
I think it should be stacks chain height.

Reply:
Seconded -- Stacks chain height.

Reply:
> Any suggestions?

A couple things:
* for names whose `renewal_deadline` becomes 0 or negative after subtracting `block`, don't include the name (since it's expired).
* for names whose `first_registered` values become negative, set them to 0.  I don't think we need to worry about what height they were registered at before the launch of Stacks 2.0 -- anyone who wants to know this information is free to spin up a Stacks 1.0 node.

Reply:
`expire_block` is not used (and we can remove it from that dump), because this property is calculated.
If we're setting the `registered at` to 0, then we would be extending all the leases.
If we're not basing on the burnchain height, I'll update the contract and add a `if` and a `legacy_expire_block` field to track the deadline set in v1.

Reply:
The reason I recommended setting `first_registered` to `0`0 is because `registered at` can't be negative in BNS -- both `registered-at` and `imported-at` are `uint` in `bns.clar`.

Reply:
`renewal_deadline` is now removed. Expired names are already not included. 

Do we want to extend name expirations by setting `first_registered` to `0`, or change the contract to support signed ints? Seems like those are the only two options?

Reply:
Yup, sounds like it.  Let's ask @diwakergupta.

Reply:
@zone117x let's do the former, and extend the expirations.

Reply:
Got word from @diwakergupta -- let's just renew them all.  Set `first_registered` to 0.

Reply:
Great, done!

Reply:
Thoughts on if we should also drop `reveal_block` and `ready_block` from the `namespaces` table? These will also always be negative and I think require a contract change to handle. @jcnelson @lgalabru

Reply:
Removed the `reveal_block` and `ready_block` from the `namespaces` table as discussed on Discord.